In [568]:
import pandas as pd
import numpy as np
from pathlib import Path

In [569]:
# Get current working path
os.getcwd()

'/Users/ianlin/Desktop/AlphaCampPractice'

<font color=white size=6><mark style=background-color:grey> Part 1 問題 </mark></font>

<font size=5> Q1. 載入資料＆簡易分析 </font>


請將上述 metadata 資料載入成為 Pandas 的 DataFrame

In [570]:
# Read _JSON
dataFile_path = os.getcwd() + '/data/meta_All_Beauty.json.gz'
print('Reading... ', dataFile_path)
meta_All_Beauty_df = pd.read_json(dataFile_path, lines=True, compression='gzip')
print('Done')

Reading...  /Users/ianlin/Desktop/AlphaCampPractice/data/meta_All_Beauty.json.gz
Done


1. 在此份檔案當中，總共有多少筆資料？多少欄位？

In [571]:
# Get the shape of the Data Frame
meta_All_Beauty_df.shape

(32892, 19)

In [572]:
meta_All_Beauty_df.head(5)

category tech1                                        description fit  \
0       []        [Loud 'N Clear Personal Sound Amplifier allows...       
1       []        [No7 Lift & Luminate Triple Action Serum 50ml ...       
2       []        [No7 Stay Perfect Foundation now stays perfect...       
3       []                                                       []       
4       []        [Lacto Calamine Skin Balance Daily Nourishing ...       

                                               title  \
0      Loud 'N Clear&trade; Personal Sound Amplifier   
1  No7 Lift &amp; Luminate Triple Action Serum 50...   
2    No7 Stay Perfect Foundation Cool Vanilla by No7   
3  Wella Koleston Perfect Hair Colour 44/44 Mediu...   
4  Lacto Calamine Skin Balance Oil control 120 ml...   

                   also_buy tech2              brand feature  \
0                        []             idea village      []   
1  [B01E7LCSL6, B008X5RVME]                               []   
2                        []                      No7      []   
3              [B0041PBXX8]                               []   
4                        []        Pirmal Healthcare      []   

                                    rank  \
0  2,938,573 in Beauty & Personal Care (   
1    872,854 in Beauty & Personal Care (   
2    956,696 in Beauty & Personal Care (   
3  1,870,258 in Beauty & Personal Care (   
4     67,701 in Beauty & Personal Care (   

                                           also_view  \
0                                                 []   
1                                                 []   
2               [B01B8BR0O8, B01B8BR0NO, B014MHXXM8]   
3                                                 []   
4  [3254895630, B007VL1D9S, B00EH9A0RI, B0773MBG4...   

                                             details    main_cat  \
0                           {'ASIN: ': '6546546450'}  All Beauty   
1  {'Shipping Weight:': '0.3 ounces (', 'ASIN: ':...  All Beauty   
2  {'Shipping Weight:': '3.5 ounces (', 'ASIN: ':...  All Beauty   
3  {'
    Item Weight: 
    ': '1.76 ounces', 'Sh...  All Beauty   
4  {'Shipping Weight:': '12 ounces (', 'ASIN: ': ...  All Beauty   

                                        similar_item date   price        asin  \
0                                                                  6546546450   
1   class="a-bordered a-horizontal-stripes  a-spa...       $44.99  7178680776   
2                                                          $28.76  7250468162   
3                                                                  7367905066   
4                                                          $12.15  7414204790   

                                            imageURL  \
0                                                 []   
1                                                 []   
2                                                 []   
3  [https://images-na.ssl-images-amazon.com/image...   
4  [https://images-na.ssl-images-amazon.com/image...   

                                     imageURLHighRes  
0                                                 []  
1                                                 []  
2                                                 []  
3  [https://images-na.ssl-images-amazon.com/image...  
4  [https://images-na.ssl-images-amazon.com/image...

In [573]:
meta_All_Beauty_df.columns

Index(['category', 'tech1', 'description', 'fit', 'title', 'also_buy', 'tech2',
       'brand', 'feature', 'rank', 'also_view', 'details', 'main_cat',
       'similar_item', 'date', 'price', 'asin', 'imageURL', 'imageURLHighRes'],
      dtype='object')

2. 在此份資料當中，擁有最多品項的前三名品牌為？

In [574]:
# Find the top-3 best seller
meta_All_Beauty_df['brand'].value_counts().head(4)

                   15673
VAGA                 252
L'Oreal Paris        113
Philips Norelco       88
Name: brand, dtype: int64

<font size=5> Q2. 清理資料＆調整格式 </font>

1. 移除資料中不必要的欄位，只留下
<mark style=background-color:grey><font color=red > 'asin' </font></mark>,
<mark style=background-color:grey><font color=red > 'brand' </font></mark>,
<mark style=background-color:grey><font color=red > 'title' </font></mark>,
<mark style=background-color:grey><font color=red > 'price' </font></mark>,
<mark style=background-color:grey><font color=red > 'rank' </font></mark>,
<mark style=background-color:grey><font color=red > 'description' </font></mark>


In [575]:
meta_All_Beauty_df[['asin', 'brand', 'title', 'price', 'rank', 'description']].head(5)

asin              brand  \
0  6546546450       idea village   
1  7178680776                      
2  7250468162                No7   
3  7367905066                      
4  7414204790  Pirmal Healthcare   

                                               title   price  \
0      Loud 'N Clear&trade; Personal Sound Amplifier           
1  No7 Lift &amp; Luminate Triple Action Serum 50...  $44.99   
2    No7 Stay Perfect Foundation Cool Vanilla by No7  $28.76   
3  Wella Koleston Perfect Hair Colour 44/44 Mediu...           
4  Lacto Calamine Skin Balance Oil control 120 ml...  $12.15   

                                    rank  \
0  2,938,573 in Beauty & Personal Care (   
1    872,854 in Beauty & Personal Care (   
2    956,696 in Beauty & Personal Care (   
3  1,870,258 in Beauty & Personal Care (   
4     67,701 in Beauty & Personal Care (   

                                         description  
0  [Loud 'N Clear Personal Sound Amplifier allows...  
1  [No7 Lift & Luminate Triple Action Serum 50ml ...  
2  [No7 Stay Perfect Foundation now stays perfect...  
3                                                 []  
4  [Lacto Calamine Skin Balance Daily Nourishing ...

2. 將銷售排名
<mark style=background-color:grey><font color=red > 'rank'</font></mark>
中的排名字串抽取出來、轉為數字

In [576]:
meta_All_Beauty_df['rank'].head(5)

0    2,938,573 in Beauty & Personal Care (
1      872,854 in Beauty & Personal Care (
2      956,696 in Beauty & Personal Care (
3    1,870,258 in Beauty & Personal Care (
4       67,701 in Beauty & Personal Care (
Name: rank, dtype: object

In [577]:
# Create a new column 'rankNum'
meta_All_Beauty_df.insert(5,'rankNum', '')

In [578]:
# Extract the number
meta_All_Beauty_df['rankNum'] = meta_All_Beauty_df['rank'].str.extract(pat = "((\d{1,3})(,\d\d\d)*)")[0]

# Remove comma (,)
meta_All_Beauty_df['rankNum'] = meta_All_Beauty_df['rankNum'].str.replace(',', '')
meta_All_Beauty_df['rankNum'] = pd.to_numeric(meta_All_Beauty_df['rankNum'])
meta_All_Beauty_df['rankNum'].head(5)

0    2938573.0
1     872854.0
2     956696.0
3    1870258.0
4      67701.0
Name: rankNum, dtype: float64

**<font size=3> 請回答： </font>**

1. 找到資料中銷售排名最高的產品是第幾名？
2. 最高銷售排名的產品為何？

In [579]:
# Find the product with the highest ranking
meta_All_Beauty_df.sort_values(by='rankNum', ascending=True)[['title', 'rankNum']].head(5)

title  rankNum
8540   Braun Clean &amp; Renew Refill Cartridges CCR,...     35.0
6268   Aquaphor Healing Ointment,Advanced Therapy Ski...     65.0
12028  Astra Platinum Double Edge Safety Razor Blades...    245.0
3606   Astra Platinum Double Edge Safety Razor Blades...    280.0
481    Panasonic Bikini Shaper and Trimmer for Women ...    457.0

<font color=white size=6><mark style=background-color:grey> Part 2 問題 </mark></font>

<font size=5> Q3. 聚合資料 </font>

pandas 載入美妝品的評價資料

In [580]:
# Read _JSON
dataFile_path = os.getcwd() + '/data/All_Beauty.json.gz'
print('Reading... ', dataFile_path)
All_Beauty_df = pd.read_json(dataFile_path, lines=True, compression='gzip')
print('Done')

Reading...  /Users/ianlin/Desktop/AlphaCampPractice/data/All_Beauty.json.gz
Done


In [581]:
All_Beauty_df.head(10)

overall  verified   reviewTime      reviewerID        asin  \
0        1      True  02 19, 2015  A1V6B6TNIC10QE  0143026860   
1        4      True  12 18, 2014  A2F5GHSXFQ0W6J  0143026860   
2        4      True  08 10, 2014  A1572GUYS7DGSR  0143026860   
3        5      True  03 11, 2013   A1PSGLFK1NSVO  0143026860   
4        5      True  12 25, 2011   A6IKXKZMTKGSC  0143026860   
5        5     False  02 26, 2010  A36NF437WZLQ9E  0143026860   
6        4     False   03 7, 2001  A10Q8NIFOVOHFV  0143026860   
7        1      True  04 10, 2017  A26PO1B2Q2G1CS  014789302X   
8        5      True   01 3, 2017   AQ812VYVTC2RJ  014789302X   
9        1      True  11 19, 2016   AMACNEW14ADMX  014789302X   

         reviewerName                                         reviewText  \
0   theodore j bigham                                              great   
1        Mary K. Byke  My  husband wanted to reading about the Negro ...   
2             David G  This book was very informative, covering all a...   
3                TamB  I am already a baseball fan and knew a bit abo...   
4          shoecanary  This was a good story of the Black leagues. I ...   
5           W. Powell  Today I gave a book about the Negro Leagues of...   
6  Robert S. Clay Jr.  The story of race relations in American histor...   
7     Jacqueline Diaz  I didn't like this product it smudged all unde...   
8  Khadijah Ali-Evans  I simply love the product. I appreciate print ...   
9          rabiyaa123  it burns your eyes when u put it on  and very ...   

                                             summary  unixReviewTime vote  \
0                                           One Star      1424304000  NaN   
1  ... to reading about the Negro Baseball and th...      1418860800  NaN   
2                                     Worth the Read      1407628800  NaN   
3                                          Good Read      1362960000  NaN   
4                More than facts, a good story read!      1324771200    5   
5                         The Gift of Black Baseball      1267142400    4   
6                     Baseball, America, and racism.       983923200   14   
7                                           One Star      1491782400  NaN   
8                                         Five Stars      1483401600  NaN   
9                                i do not recommend.      1479513600    3   

  style image  
0   NaN   NaN  
1   NaN   NaN  
2   NaN   NaN  
3   NaN   NaN  
4   NaN   NaN  
5   NaN   NaN  
6   NaN   NaN  
7   NaN   NaN  
8   NaN   NaN  
9   NaN   NaN

In [582]:
# Check the shape
All_Beauty_df.shape

(371345, 12)

In [583]:
# Check the columns
All_Beauty_df.columns

Index(['overall', 'verified', 'reviewTime', 'reviewerID', 'asin',
       'reviewerName', 'reviewText', 'summary', 'unixReviewTime', 'vote',
       'style', 'image'],
      dtype='object')

In [584]:
# Keep only necessary columns
All_Beauty_df[['overall', 'verified', 'asin', 'reviewText', 'summary', 'vote']].head(10)

overall  verified        asin  \
0        1      True  0143026860   
1        4      True  0143026860   
2        4      True  0143026860   
3        5      True  0143026860   
4        5      True  0143026860   
5        5     False  0143026860   
6        4     False  0143026860   
7        1      True  014789302X   
8        5      True  014789302X   
9        1      True  014789302X   

                                          reviewText  \
0                                              great   
1  My  husband wanted to reading about the Negro ...   
2  This book was very informative, covering all a...   
3  I am already a baseball fan and knew a bit abo...   
4  This was a good story of the Black leagues. I ...   
5  Today I gave a book about the Negro Leagues of...   
6  The story of race relations in American histor...   
7  I didn't like this product it smudged all unde...   
8  I simply love the product. I appreciate print ...   
9  it burns your eyes when u put it on  and very ...   

                                             summary vote  
0                                           One Star  NaN  
1  ... to reading about the Negro Baseball and th...  NaN  
2                                     Worth the Read  NaN  
3                                          Good Read  NaN  
4                More than facts, a good story read!    5  
5                         The Gift of Black Baseball    4  
6                     Baseball, America, and racism.   14  
7                                           One Star  NaN  
8                                         Five Stars  NaN  
9                                i do not recommend.    3

**<font size=3> 請回答： </font>**

1. 找到評價數最多的產品其 asin
2. 上述產品的評價次數為？

In [585]:
# Find the one with the most reviews
All_Beauty_df['asin'].value_counts().head(5)

B000FOI48G    8672
B000GLRREU    8341
1620213982    4792
B001QY8QXM    4544
B01DKQAXC0    4246
Name: asin, dtype: int64

<mark style=background-color:grey><font color=red>
Ans: </mark>
1. B000FOI48G
2. 8672

</font>

3. 上述產品的平均評價為？

In [586]:
All_Beauty_df.groupby(by='asin')['overall'].mean().loc['B000FOI48G']

4.393450184501845

<mark style=background-color:grey><font color=red>
Ans: </mark>

3. 4.393

</font>

<font size=5> Q4. 結合兩張表格 </font>

將 Q2 取得的產品資料 DataFrame，加入 Q4 的評價資料作為欄位
1. `overall_mean` 紀錄產品的平均評分
2. `reviews_count` 紀錄評價次數

In [587]:
# Set 'asin' as the index
meta_All_Beauty_df = meta_All_Beauty_df.set_index('asin')
meta_All_Beauty_df[['brand', 'title', 'price', 'rank', 'description']].head(5)

brand  \
asin                            
6546546450       idea village   
7178680776                      
7250468162                No7   
7367905066                      
7414204790  Pirmal Healthcare   

                                                        title   price  \
asin                                                                    
6546546450      Loud 'N Clear&trade; Personal Sound Amplifier           
7178680776  No7 Lift &amp; Luminate Triple Action Serum 50...  $44.99   
7250468162    No7 Stay Perfect Foundation Cool Vanilla by No7  $28.76   
7367905066  Wella Koleston Perfect Hair Colour 44/44 Mediu...           
7414204790  Lacto Calamine Skin Balance Oil control 120 ml...  $12.15   

                                             rank  \
asin                                                
6546546450  2,938,573 in Beauty & Personal Care (   
7178680776    872,854 in Beauty & Personal Care (   
7250468162    956,696 in Beauty & Personal Care (   
7367905066  1,870,258 in Beauty & Personal Care (   
7414204790     67,701 in Beauty & Personal Care (   

                                                  description  
asin                                                           
6546546450  [Loud 'N Clear Personal Sound Amplifier allows...  
7178680776  [No7 Lift & Luminate Triple Action Serum 50ml ...  
7250468162  [No7 Stay Perfect Foundation now stays perfect...  
7367905066                                                 []  
7414204790  [Lacto Calamine Skin Balance Daily Nourishing ...

In [591]:
# overall_mean
meta_All_Beauty_df['overall_mean'] = All_Beauty_df.groupby(by='asin')['overall'].mean()
# 'reviews_count'
meta_All_Beauty_df['reviews_count'] = All_Beauty_df.groupby(by=['asin']).size()
meta_All_Beauty_df[['brand', 'title', 'price', 'rank', 'description', 'overall_mean', 'reviews_count']].head(5)

brand  \
asin                            
6546546450       idea village   
7178680776                      
7250468162                No7   
7367905066                      
7414204790  Pirmal Healthcare   

                                                        title   price  \
asin                                                                    
6546546450      Loud 'N Clear&trade; Personal Sound Amplifier           
7178680776  No7 Lift &amp; Luminate Triple Action Serum 50...  $44.99   
7250468162    No7 Stay Perfect Foundation Cool Vanilla by No7  $28.76   
7367905066  Wella Koleston Perfect Hair Colour 44/44 Mediu...           
7414204790  Lacto Calamine Skin Balance Oil control 120 ml...  $12.15   

                                             rank  \
asin                                                
6546546450  2,938,573 in Beauty & Personal Care (   
7178680776    872,854 in Beauty & Personal Care (   
7250468162    956,696 in Beauty & Personal Care (   
7367905066  1,870,258 in Beauty & Personal Care (   
7414204790     67,701 in Beauty & Personal Care (   

                                                  description  overall_mean  \
asin                                                                          
6546546450  [Loud 'N Clear Personal Sound Amplifier allows...           2.5   
7178680776  [No7 Lift & Luminate Triple Action Serum 50ml ...           3.0   
7250468162  [No7 Stay Perfect Foundation now stays perfect...           5.0   
7367905066                                                 []           5.0   
7414204790  [Lacto Calamine Skin Balance Daily Nourishing ...           4.4   

            reviews_count  
asin                       
6546546450            2.0  
7178680776            1.0  
7250468162            1.0  
7367905066            1.0  
7414204790           15.0

1. 最高銷售排名的產品其使用者平均的評分？
2. 最高銷售排名的產品其使用者評價次數？

In [592]:
meta_All_Beauty_df.sort_values(by='rankNum', ascending=True)[['title', 'rankNum', 'overall_mean', 'reviews_count']].head(5)

title  rankNum  \
asin                                                                     
B00BXPVIT8  Braun Clean &amp; Renew Refill Cartridges CCR,...     35.0   
B006IB5T4W  Aquaphor Healing Ointment,Advanced Therapy Ski...     65.0   
B00HYEPVG8  Astra Platinum Double Edge Safety Razor Blades...    245.0   
B001QY8QXM  Astra Platinum Double Edge Safety Razor Blades...    280.0   
B00005JS5C  Panasonic Bikini Shaper and Trimmer for Women ...    457.0   

            overall_mean  reviews_count  
asin                                     
B00BXPVIT8      4.846154           13.0  
B006IB5T4W      4.643369         2790.0  
B00HYEPVG8      4.588235           34.0  
B001QY8QXM      4.556118         4544.0  
B00005JS5C      4.088039         2249.0

<mark style=background-color:grey><font color=red>
Ans: </mark>
1. 4.846
2. 13

</font>